src/memkind_memtier.c

```diff
@@ -1139,7 +1138,7 @@ MEMKIND_EXPORT void *memtier_kind_realloc(memkind_t kind, void *ptr,
         decrement_alloc_size(kind->partition, old_size);
         memkind_free(kind, ptr);
         return NULL;
-    } else if (ptr == NULL) {
+    } else if ((pol == MEMTIER_POLICY_DATA_HOTNESS) && (ptr == NULL)) {
```

Comment:
Lines 1142-1161 should be under pol == MEMTIER_POLICY_DATA_HOTNESS, line 1162 should be called regardless of policy.